File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.util.*;

class Solution {
    static class File {
        String head;
        String number;
        String original;
        
        File(String head, String number, String original) {
            this.head = head;
            this.number = number;
            this.original = original;
        }
    }
    public String[] solution(String[] files) {
        List<File> newFiles = new ArrayList<>();
        for (String file : files) {
            int i = 0;
            while (i < file.length() && !Character.isDigit(file.charAt(i))) {
                i++;
            }
            String head = file.substring(0, i);
            
            int j = i;
            while (j < file.length() && Character.isDigit(file.charAt(j))) {
                j++;
            }
            String number = file.substring(i, j);
            File newFile = new File(head, number, file);
            newFiles.add(newFile);
        }
        
        Collections.sort(newFiles, (f1, f2) -> {
            // head 비교
            int headCompare = f1.head.toLowerCase().compareTo(f2.head.toLowerCase());
            if (headCompare != 0) return headCompare;
        
            // number 비교
            int num1 = Integer.parseInt(f1.number);
            int num2 = Integer.parseInt(f2.number);
            return num1 - num2;
        });
            
        String[] answer = new String[files.length];
        int i = 0;
        for (File file : newFiles) {
            answer[i] = file.original;
            i++;
        }
        
        return answer;
    }
}

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
def solution(files):
    answer = []
    files_list = []
    for file in files:
        head = ''
        number = ''
        tail = ''
        i = 0

        while i < len(file) and not file[i].isdigit():
            head += file[i]
            i += 1
        
        while i < len(file) and file[i].isdigit():
            number += file[i]
            i += 1

        tail = file[i:]
        number = int(number)
        files_list.append([head, number, tail, file])
        
    files_list.sort(key=lambda x : (x[0].lower(), x[1]))
    for file in files_list:
        answer.append(file[3])
                
    return answer